File: src/auditchain/schemas/ingestion.py
"""Pydantic schemas for the company data ingestion pipeline.

These models define the request/response contracts and internal data structures
for the ingestion subsystem, which handles downloading SEC filings, parsing
XBRL data, and embedding textual disclosures for a given company.
"""

from datetime import datetime
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field, field_validator
import re


class IngestionStage(str, Enum):
    """Ordered stages of the ingestion pipeline."""

    VALIDATE = "validate"
    DOWNLOAD_FACTS = "download_facts"
    DOWNLOAD_FILINGS = "download_filings"
    PARSE_XBRL = "parse_xbrl"
    EMBED_TEXT = "embed_text"


class IngestionStatus(str, Enum):
    """Lifecycle status of an ingestion run."""

    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class IngestionRun(BaseModel):
    """Representation of an ingestion run record (mirrors ingestion_runs table)."""

    id: str
    cik: str
    status: IngestionStatus
    current_stage: Optional[IngestionStage] = None
    stages_completed: list[str] = Field(default_factory=list)
    started_at: datetime
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None
    filings_count: Optional[int] = None
    chunks_generated: Optional[int] = None
    financial_items_extracted: Optional[int] = None
    is_update: bool = False

    class Config:
        from_attributes = True


class CreateIngestionRequest(BaseModel):
    """Request payload to start a new ingestion pipeline."""

    cik: str = Field(
        ...,
        description="SEC Central Index Key — exactly 10 digits, zero-padded.",
        examples=["0000320193"],
    )
    force_update: bool = Field(
        default=False,
        description="If True, allows re-ingestion for a company that already exists in the database.",
    )

    @field_validator("cik")
    @classmethod
    def validate_cik(cls, v: str) -> str:
        if not re.match(r"^\d{10}$", v):
            raise ValueError("CIK must be exactly 10 digits (zero-padded).")
        return v


class CreateIngestionResponse(BaseModel):
    """Response returned after successfully initiating an ingestion pipeline."""

    ingestion_id: str
    status: str
    message: str
    stream_url: str

File: frontend/src/types/ingestion.ts
// ───────────────────────────────────────────────────────────
// Ingestion Pipeline Types
// Mirrors Python schemas in:
//   src/auditchain/schemas/ingestion.py
//   src/auditchain/api/events/ingestion_schemas.py
// ───────────────────────────────────────────────────────────

export type IngestionStage =
  | "validate"
  | "download_facts"
  | "download_filings"
  | "parse_xbrl"
  | "embed_text";

export type IngestionStatus = "running" | "completed" | "failed";

export interface IngestionRun {
  id: string;
  cik: string;
  status: IngestionStatus;
  current_stage: IngestionStage | null;
  stages_completed: string[];
  started_at: string;
  completed_at: string | null;
  error_message: string | null;
  filings_count: number | null;
  chunks_generated: number | null;
  financial_items_extracted: number | null;
  is_update: boolean;
}

export interface CreateIngestionRequest {
  cik: string;
  force_update?: boolean;
}

export interface CreateIngestionResponse {
  ingestion_id: string;
  status: string;
  message: string;
  stream_url: string;
}

// ───────────────────────────────────────────────────────────
// SSE Events
// ───────────────────────────────────────────────────────────

export type IngestionEventType =
  | "ingestion_started"
  | "stage_started"
  | "stage_progress"
  | "stage_completed"
  | "ingestion_completed"
  | "ingestion_failed";

export interface IngestionEventBase {
  event_type: IngestionEventType;
  ingestion_id: string;
  timestamp: string;
  elapsed_seconds: number;
}

export interface IngestionStartedEvent extends IngestionEventBase {
  event_type: "ingestion_started";
  cik: string;
  company_name: string | null;
  is_update: boolean;
}

export interface StageStartedEvent extends IngestionEventBase {
  event_type: "stage_started";
  stage: IngestionStage;
  description: string;
}

export interface StageProgressEvent extends IngestionEventBase {
  event_type: "stage_progress";
  stage: IngestionStage;
  current: number;
  total: number;
  message: string;
}

export interface StageCompletedEvent extends IngestionEventBase {
  event_type: "stage_completed";
  stage: IngestionStage;
  duration_seconds: number;
  summary: Record<string, unknown>;
}

export interface IngestionCompletedEvent extends IngestionEventBase {
  event_type: "ingestion_completed";
  cik: string;
  company_name: string;
  ticker: string | null;
  filings_count: number;
  chunks_generated: number;
  financial_items_extracted: number;
  total_duration_seconds: number;
}

export interface IngestionFailedEvent extends IngestionEventBase {
  event_type: "ingestion_failed";
  failed_stage: string;
  error_message: string;
}

export type IngestionEvent =
  | IngestionStartedEvent
  | StageStartedEvent
  | StageProgressEvent
  | StageCompletedEvent
  | IngestionCompletedEvent
  | IngestionFailedEvent;
